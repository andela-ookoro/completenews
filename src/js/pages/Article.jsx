import React from 'react';
import PropTypes from 'prop-types';
import { ShareButtons, generateShareIcon } from 'react-share';
import * as Utilties from '../utilities/utilities';
import firebase from '../utilities/firebase';
import Notification from '../action/notifyAction';
import FavouriteAction from '../action/favouriteAction';

const {
  FacebookShareButton,
  GooglePlusShareButton,
  LinkedinShareButton,
  TwitterShareButton,
} = ShareButtons;

const FacebookIcon = generateShareIcon('facebook');
const TwitterIcon = generateShareIcon('twitter');
const GooglePlusIcon = generateShareIcon('google');
const LinkedinIcon = generateShareIcon('linkedin');

/**
 * @FileOverview A class that renders Article
 * and emit a change.
 *  @extends React.Component
 * @Author okwudiri.okoro@andela.com (Okoro Celestine)
 */
class Article extends React.Component {
  /** Create Articles object  */
  constructor() {
    super();

     /**
     * Add an article to the user favourite list and update the favourite count
     * @param {object} e The object that trigger the event
     * @return {null} Return no value.
    */
    this.addArticle = ((e) => {
      const index = e.target.value;
      const articles = JSON.parse(localStorage.getItem('articles'));
      const article = articles[index];
      article.source = this.props.source;

      // get user email from local storage
      let userEmail = JSON.parse(localStorage.getItem('userProfile'))
        .email.toString().replace('.', '_');
      userEmail = userEmail.substring(0, userEmail.indexOf('@'));
      const FavouriteAddress = `/user/${userEmail}/favourite`;
      const FavouriteRef = firebase.database().ref(FavouriteAddress);

      // add article to user favourite list in firebase
      FavouriteRef.push(article)
        .then(() => {
          Notification('Headlines has been successfully added to favourites.');
          FavouriteAction(1);

          // hide the button
          const buttonID = `btnAddToFav${index}`;
          document.getElementById(buttonID).style.display = 'none';
        })
        .catch((err) => {
          Notification(`Error occurred, ${err}`);
        });
    });

    /**
     * Add an article to the user favourite list and update the favourite count
     * @param {object} e The object that trigger the event
     * @return {null} Return no value.
    */
    this.removeArticle = ((e) => {
      // get the key of the article in firebase
      const articleKey = e.target.value;
      // get the id of the article container
      const divId = `artcile${e.target.id}`;

      // get user email from local storage
      let userEmail = JSON.parse(localStorage.getItem('userProfile'))
        .email.toString().replace('.', '_');
      userEmail = userEmail.substring(0, userEmail.indexOf('@'));

      // create the article ref
      const articleAddress = `/user/${userEmail}/favourite/${articleKey}`;
      const artcileRef = firebase.database().ref(articleAddress);
      // remove the article from firebase
      artcileRef.remove()
      .then(() => {
        Notification('Article has been successfully removed from favourites.');
        FavouriteAction(-1);

        // hide the article container
        document.getElementById(divId).style.display = 'none';
      })
      .catch((err) => {
        Notification(`Error occurred, ${err}`);
      });
    });

     /**
     * convert article time to server time zone
     * @param {object} time The object that trigger the event
     * @return {string} Return the date in server time zone.
    */
    this.toServertime = ((time) => {
      let currentTime = new Date(time).toString();
      currentTime = currentTime.substring(0, currentTime.indexOf('G'));
      return currentTime;
    });
  }

  /**
   * Render the component content
   * @return {null} Return no value.
  */
  render() {
    // create the add favourite button ID
    const addFavBtnID = `btnAddToFav${this.props.id}`;
    const divID = `artcile${this.props.id}`;

    return (
      <div className="col s12 m12 l12 shadow  hoverable" id={divID}>
        <div className="article-content">
          <h6 className="header">
            <span className="articleHeader">
              {(this.props.author) ? `${this.props.author}:  ` : ''}
              {this.props.title}
            </span>
          </h6>
          <div className="card horizontal">
            <div className="card-image">
              {(this.props.urlToImage) ?
                <img
                  className="imgStyle" src={this.props.urlToImage}
                  alt="No news image"
                />
                :
                <img
                  src="https://placehold.it/800x400?text=CompleteNEWS"
                  alt="Image"
                  className="imgStyle"
                />
              }
            </div>
            <div className="card-stacked">
              <div className="card-content">
                <p className="paragraphstyle">
                  {Utilties.replaceLinks(this.props.description)}
                </p>
              </div>
              <div className="card-action">
                <p>
                  {
                    (this.props.publishedAt)
                    ?
                    `Published on ${this.toServertime(this.props.publishedAt)}`
                    :
                    ''
                  }
                </p>
                <a
                  href={this.props.url} target="_blank" rel="noopener noreferrer"
                >
                  {(this.props.url) ? `Read on  ${this.props.source} ` : ''}</a>
                <button
                  className="waves-effect waves-light"
                  value={this.props.id} onClick={this.props.scrape}
                  title="view full story"
                > Read more
                </button>
                {(this.props.isAuth) ?
                  <button
                    id={addFavBtnID}
                    value={this.props.id} onClick={this.addArticle}
                    className="waves-effect waves-light"
                    title="Add to favourite"
                  >
                  Add to favourite
                </button>
                :
                ''
                }
                {(this.props.firebaseKey) ?
                  <button
                    id={this.props.id}
                    value={this.props.firebaseKey} onClick={this.removeArticle}
                    className="waves-effect waves-light"
                    title="Add to favourite"
                  >
                  Delete favourite
                </button>
                :
                ''
                }
                <div className="col m12" style={{ paddingLeft: '0px', paddingTop: '10px' }}>
                  <div className="col m2" style={{ paddingLeft: '0px' }}>
                    <FacebookShareButton
                      url={this.props.url} title={this.props.title}
                    >
                      <FacebookIcon size={32} round={true} />
                    </FacebookShareButton>
                  </div>
                  <div className="col m2" style={{ paddingLeft: '0px' }}>
                    <LinkedinShareButton
                      url={this.props.url} title={this.props.title}
                    >
                      <LinkedinIcon size={32} round={true} />
                    </LinkedinShareButton>
                  </div>
                  <div className="col m2" style={{ paddingLeft: '0px' }}>
                    <TwitterShareButton
                      url={this.props.url} title={this.props.title}
                    >
                      <TwitterIcon size={32} round={true} />
                    </TwitterShareButton>
                  </div>
                  <div className="col m2" style={{ paddingLeft: '0px' }}>
                    <GooglePlusShareButton
                      url={this.props.url} title={this.props.title}
                    >
                      <GooglePlusIcon size={32} round={true} />
                    </GooglePlusShareButton>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>

    );
  }
}


Article.propTypes = {
  id: PropTypes.number.isRequired,
  author: PropTypes.string,
  title: PropTypes.string,
  urlToImage: PropTypes.string,
  description: PropTypes.string,
  publishedAt: PropTypes.string,
  url: PropTypes.string,
  source: PropTypes.string.isRequired,
  isAuth: PropTypes.bool.isRequired,
  scrape: PropTypes.func.isRequired,
  firebaseKey: PropTypes.string,
};

Article.defaultProps = {
  author: '',
  title: '',
  urlToImage: '',
  description: '',
  publishedAt: '',
  url: '',
  firebaseKey: ''
};

export default Article;
