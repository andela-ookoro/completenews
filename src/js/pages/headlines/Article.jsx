import React from 'react';
import PropTypes from 'prop-types';
import * as firebase from 'firebase';
// class to show article
const config = {
  apiKey: 'AIzaSyBRVO8XJcSp9EaxRRlsH5DKMILYygEUL3Y',
  authDomain: 'completenews-3a553.firebaseapp.com',
  databaseURL: 'https://completenews-3a553.firebaseio.com',
  projectId: 'completenews-3a553',
  storageBucket: 'completenews-3a553.appspot.com',
  messagingSenderId: '21466984743',
};
firebase.initializeApp(config);

class Article extends React.Component {
  constructor() {
    super();
    this.addArticle = this.addArticle.bind(this);
  }

  addArticle(e) {
    const index = e.target.value;
    const articles = JSON.parse(localStorage.articles);
    const artilce = articles[index];
    const timestamp = new Date().valueOf();
    firebase.database().ref('/article').child(timestamp).set(artilce);
    console.log(articles[index]);
    console.log(timestamp);
  }

  render() {
    const imgStyle = {
      width: '80%',
      height: '40%',
    };
    const paragraphstyle = {
      width: '80%',
    };
    return (
      <div className="col s12 m7 l12  hoverable">
        <h6 className="header">
          <span className={paragraphstyle} >
            {(this.props.author) ? this.props.author + ': ': ''} {this.props.title}
          </span>
        </h6>
        <div className="card horizontal">
          <div className="card-image">
            <img className={imgStyle} src={this.props.urlToImage} alt="No news image" />
          </div>
          <div className="card-stacked">
            <div className="card-content">
              <p className={paragraphstyle}>{this.props.description}</p>
            </div>
            <div className="card-action">
              <span>{(this.props.publishedAt) ? 'Published on ' + this.props.publishedAt : ''} </span>
              <a href={this.props.url} target="_blank" rel="noopener noreferrer">{(this.props.url) ? 'Read on site ' : ''}</a>
              <button value={this.props.id} onClick={this.addArticle} > Add to db</button>
            </div>
          </div>
        </div> 
      </div>

    );
  }
}


Article.propTypes = {
  id: PropTypes.number.isRequired,
  author: PropTypes.string,
  title: PropTypes.string,
  urlToImage: PropTypes.string,
  description: PropTypes.string,
  publishedAt: PropTypes.string,
  url: PropTypes.string,
};

export default Article;
