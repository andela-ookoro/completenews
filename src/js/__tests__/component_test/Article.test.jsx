import React from 'react';
import { mount } from 'enzyme';
import Renderer from 'react-test-renderer';
import Article from 'Article';

test('Component render the article template', () => {
  const article = {
    author: 'celestine',
    title: 'My story',
    urlToImage: 'test.jpg',
    publishedAt: '24/12/24',
    isAuth: true,
    url: 'http://localhost',
  };
  const i = 1;
  const source = 'BBC';
  const isAuth1 = true;
  const component = Renderer.create(
    <Article
      key={i} id={i} author={article.author} title={article.title}
      urlToImage={article.urlToImage} description={article.description}
      publishedAt={article.publishedAt} url={article.url} source={source}
      isAuth={isAuth1}
    />,
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('Property of each control should be same with props variable passed',
  () => {
    const article = {
      author: 'celestine',
      title: 'My story',
      urlToImage: 'test.jpg',
      publishedAt: '24/12/24',
      url: 'http://localhost',
    };
    const i = 1;
    const source = 'BBC';
    const isAuth = true;
    const wrapper = mount(
      <Article
        key={i} id={i} author={article.author} title={article.title}
        urlToImage={article.urlToImage} description={article.description}
        publishedAt={article.publishedAt} url={article.url} source={source}
        isAuth={isAuth}
      />,
    );
    const heading = wrapper.find('h6');
    expect(heading.text()).toEqual(`${article.author}:   ${article.title}`);
    const img = wrapper.find('img').props('urlToImage');
    expect(img.src).toEqual(article.urlToImage);
    const sourceURl = wrapper.find('a').props('url');
    expect(sourceURl.href).toEqual(article.url);
    expect(sourceURl.children).toEqual(`Read on  ${source} `);
    expect(sourceURl.target).toEqual('_blank');
    const button = wrapper.find('button').props('className');
    const className = button.className;
    expect(!className.includes('disabled')).toEqual(true);
  // expect(wrapper.getDOMNode()).to.have.property('card-image');
  });

test('Add to favourite button should be disabled for anonymous users', () => {
  const article = {
    author: 'celestine',
    title: 'My story',
    urlToImage: 'test.jpg',
    publishedAt: '24/12/24',
    url: 'http://localhost',
  };
  const i = 1;
  const source = 'BBC';
  const isAuth = false;
  const wrapper = mount(
    <Article
      key={i} id={i} author={article.author} title={article.title}
      urlToImage={article.urlToImage} description={article.description}
      publishedAt={article.publishedAt} url={article.url} source={source}
      isAuth={isAuth}
    />,
  );
  const button = wrapper.find('button').props('className');
  const className = button.className;
  expect(className.includes('disabled')).toEqual(true);
});

