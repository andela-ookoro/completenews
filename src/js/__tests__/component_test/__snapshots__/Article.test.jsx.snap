// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Article component rendering dom should render content as describe in the component 1`] = `
<div
  className="col s12 m12 l12  hoverable"
>
  <div
    className="article-content"
  >
    <h6
      className="header"
    >
      <span
        className="paragraphstyle"
      >
        Abhimanyu Ghoshal:  
        Facebook launches Express Wi-Fi in India to bre
      </span>
    </h6>
    <div
      className="card horizontal"
    >
      <div
        className="card-image"
      >
        <img
          alt="No news image"
          className="imgStyle"
          src="https: //cdn2.tnwcdn.com/wp-content/blogs.dir/1/files/2017/05/Facebook-Express-Wi-Fi.jpg"
        />
      </div>
      <div
        className="card-stacked"
      >
        <div
          className="card-content"
        >
          <p
            className="paragraphstyle"
          >
            After a failed effort to offer free internet acces.
          </p>
        </div>
        <div
          className="card-action"
        >
          <p>
            Published on 
          </p>
          <a
            href="https: //thenextweb.com/facebook/2017/05/04/facebook-launches-express-wi-fi-in-india-to-bring-rural-areas-online/"
            rel="noopener noreferrer"
            target="_blank"
          >
            Read on  bbc 
          </a>
          <button
            className="waves-effect waves-light"
            onClick={[Function]}
            title="view full story"
            value={0}
          >
             Read more
          </button>
          <button
            className="waves-effect waves-light"
            id="btnAddToFav"
            onClick={[Function]}
            title="Add to favourite"
            value={0}
          >
            Add to favourite
          </button>
          <div
            className="col m12"
            style={
              Object {
                "paddingLeft": "0px",
                "paddingTop": "10px",
              }
            }
          >
            <div
              className="col m2"
              style={
                Object {
                  "paddingLeft": "0px",
                }
              }
            >
              <div
                className="SocialMediaShareButton SocialMediaShareButton--facebook"
                onClick={[Function]}
                style={Object {}}
              >
                <div
                  style={
                    Object {
                      "height": 32,
                      "width": 32,
                    }
                  }
                >
                  <svg
                    className="social-icon social-icon--facebook "
                    fill="white"
                    height={32}
                    viewBox="0 0 64 64"
                    width={32}
                  >
                    <g>
                      <circle
                        cx="32"
                        cy="32"
                        fill="#3b5998"
                        r="31"
                        style={Object {}}
                      />
                    </g>
                    <g>
                      <path
                        d="M34.1,47V33.3h4.6l0.7-5.3h-5.3v-3.4c0-1.5,0.4-2.6,2.6-2.6l2.8,0v-4.8c-0.5-0.1-2.2-0.2-4.1-0.2 c-4.1,0-6.9,2.5-6.9,7V28H24v5.3h4.6V47H34.1z"
                      />
                    </g>
                  </svg>
                </div>
              </div>
            </div>
            <div
              className="col m2"
              style={
                Object {
                  "paddingLeft": "0px",
                }
              }
            >
              <div
                className="SocialMediaShareButton SocialMediaShareButton--linkedin"
                onClick={[Function]}
                style={Object {}}
              >
                <div
                  style={
                    Object {
                      "height": 32,
                      "width": 32,
                    }
                  }
                >
                  <svg
                    className="social-icon social-icon--linkedin "
                    fill="white"
                    height={32}
                    viewBox="0 0 64 64"
                    width={32}
                  >
                    <g>
                      <circle
                        cx="32"
                        cy="32"
                        fill="#007fb1"
                        r="31"
                        style={Object {}}
                      />
                    </g>
                    <g>
                      <path
                        d="M20.4,44h5.4V26.6h-5.4V44z M23.1,18c-1.7,0-3.1,1.4-3.1,3.1c0,1.7,1.4,3.1,3.1,3.1 c1.7,0,3.1-1.4,3.1-3.1C26.2,19.4,24.8,18,23.1,18z M39.5,26.2c-2.6,0-4.4,1.4-5.1,2.8h-0.1v-2.4h-5.2V44h5.4v-8.6 c0-2.3,0.4-4.5,3.2-4.5c2.8,0,2.8,2.6,2.8,4.6V44H46v-9.5C46,29.8,45,26.2,39.5,26.2z"
                      />
                    </g>
                  </svg>
                </div>
              </div>
            </div>
            <div
              className="col m2"
              style={
                Object {
                  "paddingLeft": "0px",
                }
              }
            >
              <div
                className="SocialMediaShareButton SocialMediaShareButton--twitter"
                onClick={[Function]}
                style={Object {}}
              >
                <div
                  style={
                    Object {
                      "height": 32,
                      "width": 32,
                    }
                  }
                >
                  <svg
                    className="social-icon social-icon--twitter "
                    fill="white"
                    height={32}
                    viewBox="0 0 64 64"
                    width={32}
                  >
                    <g>
                      <circle
                        cx="32"
                        cy="32"
                        fill="#00aced"
                        r="31"
                        style={Object {}}
                      />
                    </g>
                    <g>
                      <path
                        d="M48,22.1c-1.2,0.5-2.4,0.9-3.8,1c1.4-0.8,2.4-2.1,2.9-3.6c-1.3,0.8-2.7,1.3-4.2,1.6 C41.7,19.8,40,19,38.2,19c-3.6,0-6.6,2.9-6.6,6.6c0,0.5,0.1,1,0.2,1.5c-5.5-0.3-10.3-2.9-13.5-6.9c-0.6,1-0.9,2.1-0.9,3.3 c0,2.3,1.2,4.3,2.9,5.5c-1.1,0-2.1-0.3-3-0.8c0,0,0,0.1,0,0.1c0,3.2,2.3,5.8,5.3,6.4c-0.6,0.1-1.1,0.2-1.7,0.2c-0.4,0-0.8,0-1.2-0.1 c0.8,2.6,3.3,4.5,6.1,4.6c-2.2,1.8-5.1,2.8-8.2,2.8c-0.5,0-1.1,0-1.6-0.1c2.9,1.9,6.4,2.9,10.1,2.9c12.1,0,18.7-10,18.7-18.7 c0-0.3,0-0.6,0-0.8C46,24.5,47.1,23.4,48,22.1z"
                      />
                    </g>
                  </svg>
                </div>
              </div>
            </div>
            <div
              className="col m2"
              style={
                Object {
                  "paddingLeft": "0px",
                }
              }
            >
              <div
                className="SocialMediaShareButton SocialMediaShareButton--googlePlus"
                onClick={[Function]}
                style={Object {}}
              >
                <div
                  style={
                    Object {
                      "height": 32,
                      "width": 32,
                    }
                  }
                >
                  <svg
                    className="social-icon social-icon--google "
                    fill="white"
                    height={32}
                    viewBox="0 0 64 64"
                    width={32}
                  >
                    <g>
                      <circle
                        cx="32"
                        cy="32"
                        fill="#dd4b39"
                        r="31"
                        style={Object {}}
                      />
                    </g>
                    <g>
                      <path
                        d="M25.3,30.1v3.8h6.3c-0.3,1.6-1.9,4.8-6.3,4.8c-3.8,0-6.9-3.1-6.9-7s3.1-7,6.9-7c2.2,0,3.6,0.9,4.4,1.7l3-2.9c-1.9-1.8-4.4-2.9-7.4-2.9c-6.1,0-11.1,5-11.1,11.1s5,11.1,11.1,11.1c6.4,0,10.7-4.5,10.7-10.9c0-0.7-0.1-1.3-0.2-1.8H25.3L25.3,30.1z M49.8,28.9h-3.2v-3.2h-3.2v3.2h-3.2v3.2h3.2v3.2h3.2v-3.2h3.2"
                      />
                    </g>
                  </svg>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
